xbmc/platform/linux/LogindUtils.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "LogindUtils.h"
+#include "SessionUtils.h"
+
+#include "platform/linux/DBusMessage.h"
+#include "utils/log.h"
+
+#include <systemd/sd-login.h>
+#include <sys/sysmacros.h>
+
+uint32_t CSessionUtils::Open(std::string path, int flags)
+{
+  std::string sessionPath = CLogindUtils::GetSessionPath();
+
+  if (sessionPath.empty())
+  {
+    return open(path.c_str(), O_RDWR | flags);
+  }
+
+  struct stat st;
+  auto ret = stat(path.c_str(), &st);
+  if (ret < 0)
+  {
+    CLog::Log(LOGERROR, "logind: stat failed for: {}", path);
+    return -1;
+  }
+
+  if (!S_ISCHR(st.st_mode))
+  {
+    CLog::Log(LOGERROR, "logind: invalid device passed");
+    return -1;
+  }
+
+  auto fd = CLogindUtils::TakeDevice(sessionPath, major(st.st_rdev), minor(st.st_rdev));
+  if (fd < 0)
+  {
+    return fd;
+  }
+
+  auto fl = fcntl(fd, F_GETFL);
+  if (fl < 0)
+  {
+    CLog::Log(LOGERROR, "logind: F_GETFL failed");
+    close(fd);
+    CLogindUtils::ReleaseDevice(sessionPath, major(st.st_rdev), minor(st.st_rdev));
+    return -1;
+  }
+
+  if (flags & O_NONBLOCK)
```

Comment:
You can put `F_GETFL`/`F_SETFL` into the `if`

Reply:
???

Reply:
The flags you set with `F_SETFL` only ever change from the previous state if flags has `O_NONBLOCK` set. So if it isn't set, there's no point in calling `fcntl` at all.